x/lightclient/keeper/rollback.go

```diff
@@ -0,0 +1,131 @@
+package keeper
+
+import (
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/ibc-go/v7/modules/core/exported"
+
+	clienttypes "github.com/cosmos/ibc-go/v7/modules/core/02-client/types"
+	commitmenttypes "github.com/cosmos/ibc-go/v7/modules/core/23-commitment/types"
+
+	ibctm "github.com/cosmos/ibc-go/v7/modules/light-clients/07-tendermint"
+)
+
+func (hook rollappHook) OnHardFork(ctx sdk.Context, rollappId string, fraudHeight uint64) {
+	hook.k.RollbackCanonicalClient(ctx, rollappId, fraudHeight)
+}
+
+func (k Keeper) RollbackCanonicalClient(ctx sdk.Context, rollappId string, fraudHeight uint64) {
+	client, found := k.GetCanonicalClient(ctx, rollappId)
+	if !found {
+		k.Logger(ctx).Error("Canonical client not found", "rollappId", rollappId)
+		return
+	}
+	cs := k.ibcClientKeeper.ClientStore(ctx, client)
+
+	// iterate over all consensus states and metadata in the client store
+	IterateConsensusStateDescending(cs, func(h exported.Height) bool {
+		// iterate until we pass the fraud height
+		if h.GetRevisionHeight() < fraudHeight {
+			return true
+		}
+
+		// delete consensus state and metadata
+		deleteConsensusState(cs, h)
+		deleteConsensusMetadata(cs, h)
+
+		return false
+	})
+
+	// clean the optimistic updates valset
+	err := k.PruneSignersAbove(ctx, client, fraudHeight-1)
+	if err != nil {
+		k.Logger(ctx).Error("Failed to prune signers", "err", err)
+	}
+
+	// marks that hard fork is in progress
+	k.setHardForkInProgress(ctx, rollappId)
+
+	// freeze the client
+	// it will be released after the hardfork is resolved (on the next state update)
+	k.freezeClient(cs, fraudHeight)
+}
+
+// set latest IBC consensus state nextValHash to the current proposing sequencer.
+func (k Keeper) ResolveHardFork(ctx sdk.Context, rollappID string) {
```

Comment:
We're losing the flow of updating the client based on every state update first two blocks

Reply:
not sure wdym

Reply:
why we losing this flow? not sure I follow
we'll `ResolveHardFork` on the next state update

Reply:
yeah but that's only HF
It was supposed to be every state info, not just after HF

Reply:
didn't see it in the ADR
as discussed we can do if needed
in separate issue